Add ShardedTensorMetadata for ShardedTensor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61683**

This PR adds a consolidated metadata field (ShardedTensorMetadata)
which has all the necessary global metadata for a ShardedTensor.

Differential Revision: [D29703719](https://our.internmc.facebook.com/intern/diff/D29703719/)

pritamdamania committed:
Add ShardedTensorMetadata for ShardedTensor.

This PR adds a consolidated metadata field (ShardedTensorMetadata)
which has all the necessary global metadata for a ShardedTensor.

Differential Revision: [D29703719](https://our.internmc.facebook.com/intern/diff/D29703719/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -35,6 +35,25 @@ class Shard(object):
+    shards_metadata: List[ShardMetadata]
+
+    # Size of each dim of the overall Tensor.
+    overall_dims: torch.Size
```
since this is already a sharded tensor metadata, should we just call this as `size` to match with other fields?

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -337,12 +383,12 @@ def sharding_spec(self) -> ShardingSpec:
-    def sharding_metadata(self) -> List[ShardMetadata]:
+    def metadata(self) -> ShardedTensorMetadata:
         """
         Returns a list of :class:`ShardeMetadata` objects corresponding to the
```
nit: change the doc as well

wanchaol reviewed (approved):
looks great! only have some nits, thanks!

zzzwen commented:
Two questions:
1.  ShardedTensor requires Progress Group. Will it be an issue to force a PG even though the model only using RPC?
2. On the same note, shall we be exposing RPC (even if it is not _init_ or called) to users that only care about Process Group?

pritamdamania87 commented:
> 3 questions:
> 
> 1. ShardedTensor requires Progress Group. Will it be an issue to force a PG even though the model only using RPC w/ zeus?
> 2. On the same note, shall we be exposing RPC (even if it is not _init_ or called) to users that only care about Process Group?
> 3. Do we need the `global` metadata for the ShardedTensor to function?  For checkpoint, it will need to run a collective anyways, and the global view can be deduced.  My concern is that we add this extra field, without a necessity, it might become only a maintenance burden.

@zzzwen 

1. ShardedTensor initialization is designed to be SPMD style. So yes for initialization you will always need PG, but after initialization, you can either use RPC/PG for distributed computations involving the shards. Do we have a use case where we want to use this API without initializing PG?
2. We don't expose RPC to users that only care about ProcessGroup. If the user has initialized RPC, we initialize remote shards as RRefs, otherwise we don't.
3. The ShardedTensorMetadata is the only field that has information like `requires_grad`, `pin_memory` etc., don't think there are other fields from which this information can be deduced.

pritamdamania committed:
Update on "Add ShardedTensorMetadata for ShardedTensor."


This PR adds a consolidated metadata field (ShardedTensorMetadata)
which has all the necessary global metadata for a ShardedTensor.

Differential Revision: [D29703719](https://our.internmc.facebook.com/intern/diff/D29703719/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add ShardedTensorMetadata for ShardedTensor."


This PR adds a consolidated metadata field (ShardedTensorMetadata)
which has all the necessary global metadata for a ShardedTensor.

Differential Revision: [D29703719](https://our.internmc.facebook.com/intern/diff/D29703719/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add ShardedTensorMetadata for ShardedTensor."


This PR adds a consolidated metadata field (ShardedTensorMetadata)
which has all the necessary global metadata for a ShardedTensor.

Differential Revision: [D29703719](https://our.internmc.facebook.com/intern/diff/D29703719/)

[ghstack-poisoned]
